com.hazelcast.nio.serialization.HazelcastSerializationException: java.io.UTFDataFormatException: Length check failed, maybe broken bytestream or wrong stream position

Finding the first occurence of UTFDataFormatException at Line 952 in worker.log and the last

```
grep -nr "java.io.UTFDataFormatException: Length check failed" worker.log | more
952:FATAL 2014-12-09 15:23:20,258 [hz._hzInstance_1_workers.partition-operation.thread-0] com.hazelcast.spi.impl.RemoteOperationExceptionHandler: java.io.UTFDataFormatException: Length check failed, maybe
 broken bytestream or wrong stream position
...
2335945:FATAL 2014-12-09 21:41:49,255 [hz._hzInstance_1_workers.partition-operation.thread-3] com.hazelcast.spi.OperationService: [10.144.233.155]:5701 [workers] [3.4-RC1-SNAPSHOT] java.io.UTFDataFormatEx
ception: Length check failed, maybe broken bytestream or wrong stream position
```

Full Stack Trace of UTFDataFormatException at line 953  /large1/workers/2014-12-09__15_22_46/worker-10.144.233.155-7-server

```
less +953 worker.log
FATAL 2014-12-09 15:23:20,258 [hz._hzInstance_1_workers.partition-operation.thread-0] com.hazelcast.spi.impl.RemoteOperationExceptionHandler: java.io.UTFDataFormatException: Length check failed, maybe broken bytestream or wrong stream position
com.hazelcast.nio.serialization.HazelcastSerializationException: java.io.UTFDataFormatException: Length check failed, maybe broken bytestream or wrong stream position
        at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:419)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.readObject(SerializationServiceImpl.java:315)
        at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:439)
        at com.hazelcast.partition.impl.ReplicaSyncResponse.readInternal(ReplicaSyncResponse.java:178)
        at com.hazelcast.spi.Operation.readData(Operation.java:299)
        at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:111)
        at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:39)
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.toObject(StreamSerializerAdapter.java:65)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toObject(SerializationServiceImpl.java:260)
        at com.hazelcast.spi.impl.NodeEngineImpl.toObject(NodeEngineImpl.java:186)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.loadOperation(BasicOperationService.java:633)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:616)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$1500(BasicOperationService.java:609)
        at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:561)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:466)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.processPriorityMessages(BasicOperationScheduler.java:480)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:457)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:432)
Caused by: java.io.UTFDataFormatException: Length check failed, maybe broken bytestream or wrong stream position
        at com.hazelcast.nio.UTFEncoderDecoder.readUTF0(UTFEncoderDecoder.java:505)
        at com.hazelcast.nio.UTFEncoderDecoder.readUTF(UTFEncoderDecoder.java:77)
        at com.hazelcast.nio.serialization.ByteArrayObjectDataInput.readUTF(ByteArrayObjectDataInput.java:434)
        at com.hazelcast.cache.impl.operation.CacheReplicationOperation.readInternal(CacheReplicationOperation.java:155)
        at com.hazelcast.spi.Operation.readData(Operation.java:299)
        at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:111)
        at com.hazelcast.nio.serialization.DataSerializer.read(DataSerializer.java:39)
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.read(StreamSerializerAdapter.java:44)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.readObject(SerializationServiceImpl.java:309)
        ... 16 more
```

--- a/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheReplicationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheReplicationOperation.java
@@ -130,12 +130,17 @@ public class CacheReplicationOperation extends AbstractOperation {
                 final Data key = e.getKey();
                 final CacheRecord record = e.getValue();
                 final long expirationTime = record.getExpirationTime();
+
+                // If entry is already expired we skip it
                 if (expirationTime > now) {
                     out.writeData(key);
                     out.writeObject(record);
                 }
             }
-            //empty data will terminate the iteration for read
+            // Empty data will terminate the iteration for read in case
+            // expired entries were found while serializing, since the
+            // real subCount will then be different from the one written
+            // before
             out.writeData(new DefaultData());
         }
     }
@@ -155,10 +160,14 @@ public class CacheReplicationOperation extends AbstractOperation {
             String name = in.readUTF();
             Map<Data, CacheRecord> m = new HashMap<Data, CacheRecord>(subCount);
             data.put(name, m);
-            for (int j = 0; j < subCount; j++) {
+            // subCount + 1 because of the DefaultData written as the last entry
+            // which adds another Data entry at the end of the stream!
+            for (int j = 0; j < subCount + 1; j++) {
                 Data key = in.readData();
+                // Empty data received so reading can be stopped here since
+                // since the real object subCount might be different from
+                // the number on the stream due to found expired entries
                 if (key.dataSize() == 0) {
-                    //empty data received so reading done here
                     break;
                 }
                 CacheRecord record = in.readObject();

--- a/hazelcast/src/test/java/com/hazelcast/cache/CacheSerializationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cache/CacheSerializationTest.java
@@ -1,20 +1,39 @@
 package com.hazelcast.cache;
 
+import com.hazelcast.cache.impl.CachePartitionSegment;
+import com.hazelcast.cache.impl.CacheService;
+import com.hazelcast.cache.impl.HazelcastServerCachingProvider;
+import com.hazelcast.cache.impl.operation.CacheReplicationOperation;
 import com.hazelcast.cache.impl.record.CacheRecord;
 import com.hazelcast.cache.impl.record.CacheRecordFactory;
 import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.instance.HazelcastInstanceImpl;
+import com.hazelcast.instance.HazelcastInstanceProxy;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.nio.serialization.DefaultSerializationServiceBuilder;
 import com.hazelcast.nio.serialization.SerializationService;
 import com.hazelcast.nio.serialization.SerializationServiceBuilder;
+import com.hazelcast.spi.impl.NodeEngineImpl;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
+
+import javax.cache.Cache;
+import javax.cache.CacheManager;
+import javax.cache.configuration.CompleteConfiguration;
+import javax.cache.configuration.MutableConfiguration;
+import javax.cache.spi.CachingProvider;
+import java.lang.reflect.Field;
+
+import static org.junit.Assert.assertEquals;
 
 /**
  * Serialization test class for JCache
@@ -52,11 +71,57 @@ public class CacheSerializationTest extends HazelcastTestSupport {
         service.toObject(cacheRecordData);
     }
 
+    @Test
+    public void test_CacheReplicationOperation_serialization() throws Exception {
+        TestHazelcastInstanceFactory factory = new TestHazelcastInstanceFactory(1);
+        HazelcastInstance hazelcastInstance = factory.newHazelcastInstance();
+
+        try {
+            CachingProvider provider = HazelcastServerCachingProvider.createCachingProvider(hazelcastInstance);
+            CacheManager manager = provider.getCacheManager();
+
+            CompleteConfiguration configuration = new MutableConfiguration();
+            Cache cache1 = manager.createCache("cache1", configuration);
+            Cache cache2 = manager.createCache("cache2", configuration);
+            Cache cache3 = manager.createCache("cache3", configuration);
+
+            for (int i = 0; i < 1000; i++) {
+                cache1.put("key" + i, i);
+                cache2.put("key" + i, i);
+                cache3.put("key" + i, i);
+            }
+
+            HazelcastInstanceProxy proxy = (HazelcastInstanceProxy) hazelcastInstance;
+
+            Field original = HazelcastInstanceProxy.class.getDeclaredField("original");
+            original.setAccessible(true);
+
+            HazelcastInstanceImpl impl = (HazelcastInstanceImpl) original.get(proxy);
+            NodeEngineImpl nodeEngine = impl.node.nodeEngine;
+            CacheService cacheService = nodeEngine.getService(CacheService.SERVICE_NAME);
+
+            int partitionCount = nodeEngine.getPartitionService().getPartitionCount();
+
+            for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+                CachePartitionSegment segment = cacheService.getSegment(partitionId);
+
+                CacheReplicationOperation operation = new CacheReplicationOperation(segment, 1);
+                Data serialized = service.toData(operation);
+                try {
+                    service.toObject(serialized);
+                } catch (Exception e) {
+                    throw new Exception("Partition: " + partitionId, e);
+                }
+            }
+
+        } finally {
+            factory.shutdownAll();
+        }
+    }
+
     private CacheRecord createRecord(InMemoryFormat format) {
         CacheRecordFactory factory = new CacheRecordFactory(format, service);
         return factory.newRecord(randomString());
     }
-
-
 
 }

